### ISSUE TITLE
DecoratingLogger throws Encoding::UndefinedConversionError on ASCII-8BIT encoding

### ISSUE THREAD

**Issue by nikajukic** [newrelic/newrelic-ruby-agent#863]
## Description
We're logging our requests with New Relic and had some issue with logging PayPal request/response logs. They're in ASCII-8BIT encoding.

When calling `.to_json` on a message in `NewRelic::Agent::Logging::DecoratingFormatter#escape` it raises `Encoding::UndefinedConversionError: "\xC4" from ASCII-8BIT to UTF-8`

```
message = "č"

message.encoding
=> #<Encoding:UTF-8>

message.to_json
=> "\"č\""

message.force_encoding(Encoding::ASCII-8BIT).to_json
Encoding::UndefinedConversionError: "\xC4" from ASCII-8BIT to UTF-8
from /Users/user/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/activesupport-6.1.4.1/lib/active_support/core_ext/object/json.rb:40:in `to_json'
```

This issue is similar to #842.

## Expected Behavior
The behaviour should be as it's been solved in #844 by scrubbing these characters or by forcing the UTF-8 encoding.

```
message.force_encoding(Encoding::ASCII_8BIT).force_encoding(Encoding::UTF_8).to_json
=> "\"č\""
```

## Steps to Reproduce
If a message that is passed to `NewRelic::Agent::Logging::DecoratingFormatter#escape` is in `Encoding::ASCII_8BIT` it will raise an error when calling `.to_json` on it.

```
Rails.logger.info "č"
č

Rails.logger.info "č".force_encoding(Encoding::ASCII_8BIT)
č
Encoding::UndefinedConversionError: "\xC4" from ASCII-8BIT to UTF-8
from /Users/nika/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/activesupport-6.1.4.1/lib/active_support/core_ext/object/json.rb:40:in `to_json'
```

## Your Environment
Ruby 2.6.1
Rails 6.1.4.1
New Relic - tried with 7.2.0 and current [master commit](https://github.com/newrelic/newrelic-ruby-agent/commit/a1a5fe17e1311cf6324c5513ed0c1f615466e730) (8.2.0-pre)

**Comment by kaylareopelle** [comment]
Hi @nikajukic! Thanks for bringing this to our attention. It's helpful to know the fix in #844 doesn't cover this case. 

[I have a branch](https://github.com/newrelic/newrelic-ruby-agent/pull/864) that uses another strategy to encode non-UTF-8 strings. Would you be willing to update your Gemfile to run on this branch and let me know if it resolves the issue? 

```ruby
gem 'newrelic_rpm', git: 'https://github.com/newrelic/newrelic-ruby-agent.git', branch: 'bugfix/decorating-logger-throws-EncodingUndefinedConversionError-on-ASCII-8BIT-encoding'
``` 

Thank you! 
Kayla

**Comment by nikajukic** [comment]
Hi @kaylareopelle, this solution won't work because ASCII-8BIT is still a valid encoding so `.valid_encoding?` returns true and message will never be encoded with `Encoding::UTF_8` like you intended.

```
"č".force_encoding(Encoding::ASCII_8BIT).valid_encoding?
=> true
```

Try adding this test to test/new_relic/agent/logging_test.rb
```
def test_to_replace_ascii_8bit_chars
  message = 'message with an ASCII-8BIT character'
  input = "#{message} #{"č".force_encoding(Encoding::ASCII_8BIT)}"
  expectation = "#{message} #{DecoratingFormatter::REPLACEMENT_CHAR}#{DecoratingFormatter::REPLACEMENT_CHAR}"
  logger = DecoratingLogger.new @output
  logger.info input
  assert_equal expectation, last_message['message']
end
```

This is intended behavior. It will replace ASCII-8BIT `č` with two replacement charachters because `č` consists of `"\xC4\x8D"` and each character is invalid in UTF-8 and should therefore be replaced.

**Comment by kaylareopelle** [comment]
Hi @nikajukic, thank you for the clarification. I misunderstood your earlier message and appreciate the additional information and the test case. I'll continue working on a solution that includes this test case and will let you know when a new branch is available to test.

**Comment by kaylareopelle** [comment]
Hi @nikajukic, I updated the condition to `message.encoding == Encoding::UTF_8 && message.valid_encoding?` before running `#encode` on the String. This seems to work with all the provided test cases. 

Here's a link to the diff: https://github.com/newrelic/newrelic-ruby-agent/pull/864/files#diff-e874717e708c43b82c4b577a280c3cc75f38d8aef7589a67d295b0bfa2453e62R92 

Would you be willing to give the branch a try to see if this resolves the issue?

```ruby 
gem 'newrelic_rpm', git: 'https://github.com/newrelic/newrelic-ruby-agent.git', branch: 'bugfix/decorating-logger-throws-EncodingUndefinedConversionError-on-ASCII-8BIT-encoding'
```

**Comment by kaylareopelle** [comment]
Hi @nikajukic  - I wanted to let you know a change in response to your issue is included in the newly published version 8.3.0 of the agent. We'll be deleting the branch, `bugfix/decorating-logger-throws-EncodingUndefinedConversionError-on-ASCII-8BIT-encoding` shortly.

Thanks for your contribution!
